### ISSUE TITLE
Resource Owner Password Credentials Grant and oidc provider

### ISSUE THREAD

**Issue by ll4strw** [irods/irods_client_http_api#385]
If I run the API server in oidc client mode, the server expects an oidc provider to be up-and-running to start.  This happens regardless I am planning to use the `Resource Owner Password Credentials Grant` auth scheme or the `Authorization Code Grant` auth scheme.  

The presence of an oidc provider should not be needed in the context of the `Resource Owner Password Credentials Grant` scheme of this server. Or am I wrong?
Why should an admin be forced to set up an oidc provider if the plan is to use  the `Resource Owner Password Credentials Grant` auth scheme? Besides, the presence of an oidc provider is only enforced upon booting the server. Afterwards, it is possible to shutdown the oidc provider without having the http server to complain and to continue using the API as normal with the password scheme.

Could it be an idea to have different boot requirements for these two auth schemes?

**Comment by MartinFlores751** [comment]
> If I run the API server in oidc client mode, the server expects an oidc provider to be up-and-running to start. This happens regardless I am planning to use the `Resource Owner Password Credentials Grant` auth scheme or the `Authorization Code Grant` auth scheme.
> 
> The presence of an oidc provider should not be needed in the context of the `Resource Owner Password Credentials Grant` scheme of this server. Or am I wrong?

The `Resource Owner Password Credentials Grant` forwards the provided username and password provided to the OpenID provider, in exchange for a token. Communication with a provider is still needed in this case.

See Figure 5 for a diagram of the communication: https://www.rfc-editor.org/rfc/rfc6749.html#section-4.3

If you don't want to use OpenID at all, you can remove the `openid_connect` section from your configuration to then only authenticate using your iRODS username and password.

> Why should an admin be forced to set up an oidc provider if the plan is to use the `Resource Owner Password Credentials Grant` auth scheme? Besides, the presence of an oidc provider is only enforced upon booting the server.
> 
> ...
> 
> Could it be an idea to have different boot requirements for these two auth schemes?

The idea of not starting the server when an OpenID provider cannot be found is to have the user check the configuration is correct. If the server goes down when running, it's assumed the provider will come back up, so we continue running with the existing information we have.

> ... Afterwards, it is possible to shutdown the oidc provider without having the http server to complain and to continue using the API as normal with the password scheme. ...

This is a good point. The Basic auth scheme that uses iRODS is always present and cannot currently be turned off, even if you are only using OpenID. Wonder if anyone else has any thoughts on this? @korydraughn, any thoughts?

---

Just to be sure, are you sending an `Authorization: iRODS ...` or an `Authorization: Bearer ...`?

The `Authorization: iRODS ...` should be what is used for the `Resource Owner Password Credentials Grant` for the HTTP API.

The `Authorization: Bearer ...` should be used for iRODS authorization. This does not require an OpenID Provider.

**Comment by korydraughn** [comment]
I've repurposed #38 to be about endpoints and created #387 for the Basic auth scheme.

**Comment by trel** [comment]
> The `Authorization: iRODS ...` should be what is used for the Resource Owner Password Credentials Grant for the HTTP API.
> 
> The `Authorization: Bearer ...` should be used for iRODS authorization. This does not require an OpenID Provider.

when / seen phrased like this... it's quite possible someone will get confused.

i only offer that observation.   i offer no solution/suggestion at this point.

and also, i understand why what you've posted is correct.... just looks funny to see iRODS not align with iRODS in that list of two.

**Comment by MartinFlores751** [comment]
> > The `Authorization: iRODS ...` should be what is used for the Resource Owner Password Credentials Grant for the HTTP API.
> > The `Authorization: Bearer ...` should be used for iRODS authorization. This does not require an OpenID Provider.
> 
> when / seen phrased like this... it's quite possible someone will get confused.
> 
> i only offer that observation. i offer no solution/suggestion at this point.
> 
> and also, i understand why what you've posted is correct.... just looks funny to see iRODS not align with iRODS in that list of two.

Yes, it is confusing for it to not match up.

Relevant: #363

We will be removing this grant, so the odd naming should not be around for much longer.

**Comment by trel** [comment]
> We will be removing this grant, so the odd naming should not be around for much longer.

ah, right.   carry on.

**Comment by MartinFlores751** [comment]
> Well, I did get confused 🤔 Would that also explain why updating a user passwd on the oidc provider was not affecting their authentication on the API server?

Yes, if you update your OpenID password and the change is not reflected when using an OpenID endpoint, then the incorrect `Authorization` is being used in this case.

**Comment by MartinFlores751** [comment]
Closing since client mode is now removed: #416.
